File: Algorithms/Basic/Multiplication.cpp
/*
Name:           Syed Owais Ali Chishti
Roll no:        P146011
Sec:            A
Description:    First assignment implement multiplication using divide and conquor rule.
*/

#include <iostream>
#include <string>
#include <math.h>
#include <cstdlib>
using namespace std;

int ToInt(double x)
{
   double dx = x < 0.0 ? -0.5 : 0.5;
   return static_cast<int>(x + dx);
}

int multiplyPoly(string ValA, string ValB, int split) {
	int result = 0;
	int IntB = atoi(ValB.c_str());
	for (int i = 0; i < ValB.length(); i += split) {
		int PartB = IntB % ToInt(pow(10, split));
		IntB /= ToInt(pow(10, split));
		int resultT = 0;
		int IntA = atoi(ValA.c_str());        
		for (int j = 0; j < ValA.length(); j += split) {
	        int ShiftVal = ToInt(pow(10, j + i));
            int TmpRes = PartB * (IntA % ToInt(pow(10, split))) * ShiftVal;
			resultT += TmpRes;
			cout << PartB << " x " << (IntA % ToInt(pow(10, split))) << " = " << TmpRes << "    " << endl;
            IntA /= ToInt(pow(10, split));
		}
		result += resultT;
	}
	return result;
}

int main() {
	cout << multiplyPoly("123456", "2345", 2) << endl;
	cout << 123456 * 2345;
    return 0;
}

File: Algorithms/Basic/Multiplication.java
// This is implementation of Divide and conquer Multiplication Algorithm

public class Multiplication {
    public static int toInt(char tmp) {
        return tmp - '0';
    }
    
    public static int multiplyPoly (String ValA, String ValB, int split) {
        int result = 0;

        int IntB =  Integer.parseInt(ValB);
        for (int i = 0; i < ValB.length(); i+=split) {
            int IntA =  Integer.parseInt(ValA);
            int PartB = IntB % (int)Math.pow(10,split);
            IntB /= (int)Math.pow(10,split);
            int resultT = 0;
            for (int j = 0; j < ValA.length(); j+=split) {
                int ShiftVal = (int)(Math.pow(10, j+i));
                
                int TmpRes = PartB * (IntA % (int)Math.pow(10,split)) * ShiftVal;

                resultT += TmpRes;
                System.out.println(PartB + " x " + (IntA % (int)Math.pow(10,split)) + " = " + TmpRes + "    "  + resultT );
                
                IntA /= (int)Math.pow(10,split);                
            }
            result += resultT;
        }
        return result;
    }
    
    public static int multiplySingle(String ValA, String ValB) {
        int result = 0;
        for (int i = 0; i < ValB.length(); i++) {
            int resultT = 0;
            for (int j = 0; j < ValA.length(); j++) {
                int ShiftVal = (int)(Math.pow(10, (ValA.length() - j - 1) +  (ValB.length() - i - 1)));
                int TmpRes = toInt(ValB.charAt(i)) * toInt(ValA.charAt(j)) * ShiftVal;
                resultT += TmpRes;
                System.out.println(ValB.charAt(i) + " x " + ValA.charAt(j) + " = " + TmpRes + "    "  + resultT );
            }
            result += resultT;
        }
        return result;
    }
    
    public static void main(String []args) {    
        System.out.println(multiplyPoly("13", "12", 2));
        System.out.println(13*12);

        System.out.println();
        System.out.println(multiplyPoly("1234", "432", 2));
        System.out.println(1234*432);

        System.out.println();        
        System.out.println(multiplyPoly("121334", "43123", 3));
        System.out.println(121334*43123);
        
        System.out.println();
        System.out.println(multiplyPoly("981", "1234", 2));
        System.out.println(981*1234);
                 
        System.out.println();
        System.out.println(multiplySingle("1234", "452"));
        System.out.println(1234*452);
        
        System.out.println();
        System.out.println(multiplySingle("13", "12"));
        System.out.println(13 * 12);
    }
}